File: ai-service/src/ai_service/config/subscription.py
import yaml
import os
import threading
import time
import logging

logger = logging.getLogger(__name__)

PLANS_CONFIG_PATH = "config/plans.yaml"
CONFIG_RELOAD_INTERVAL = 300  # 5 minutes

class PlanQuotas:
    def __init__(self, orders_per_day: int, tokens_per_day: int):
        self.orders_per_day = orders_per_day
        self.tokens_per_day = tokens_per_day

class PlanFeatures:
    def __init__(self, sources: list[str], pipeline: list[str], feedback_loop: bool = False):
        self.sources = sources
        self.pipeline = pipeline
        self.feedback_loop = feedback_loop

class Plan:
    def __init__(self, name: str, model: str, quotas: dict, features: dict):
        self.name = name
        self.model = model
        self.quotas = PlanQuotas(**quotas)
        self.features = PlanFeatures(**features)

class SubscriptionConfig:
    def __init__(self, plans: dict):
        self.plans = {k: Plan(name=v['name'], model=v['model'], quotas=v['quotas'], features=v['features']) for k, v in plans.items()}

_subscription_config: SubscriptionConfig = None
_subscription_config_lock = threading.RLock()

def load_subscription_config():
    global _subscription_config
    try:
        with open(PLANS_CONFIG_PATH, 'r') as f:
            data = yaml.safe_load(f)
        with _subscription_config_lock:
            _subscription_config = SubscriptionConfig(**data)
        logger.info("Subscription config loaded successfully.")
    except Exception as e:
        logger.error(f"Failed to load subscription config: {e}")
        raise

def get_subscription_config() -> SubscriptionConfig:
    with _subscription_config_lock:
        return _subscription_config

def start_subscription_config_reloader():
    def reloader_loop():
        while True:
            time.sleep(CONFIG_RELOAD_INTERVAL)
            try:
                load_subscription_config()
                logger.info("Subscription config reloaded successfully.")
            except Exception as e:
                logger.error(f"Failed to reload subscription config: {e}")

    reloader_thread = threading.Thread(target=reloader_loop, daemon=True)
    reloader_thread.start()

# Initial load
load_subscription_config()

File: backend/internal/config/subscription.go
package config

import (
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"

	"gopkg.in/yaml.v3"
)

const (
	plansConfigPath     = "config/plans.yaml"
	configReloadInterval = 5 * time.Minute
)

type Plan struct {
	Name     string             `yaml:"name"`
	Model    string             `yaml:"model"`
	Quotas   PlanQuotas         `yaml:"quotas"`
	Features PlanFeatures       `yaml:"features"`
}

type PlanQuotas struct {
	OrdersPerDay  int `yaml:"orders_per_day"`
	TokensPerDay  int `yaml:"tokens_per_day"`
}

type PlanFeatures struct {
	Sources  []string `yaml:"sources"`
	Pipeline []string `yaml:"pipeline"`
	FeedbackLoop bool `yaml:"feedback_loop"`
}

type SubscriptionConfig struct {
	Plans map[string]Plan `yaml:"plans"`
}

var (
	subscriptionConfig     *SubscriptionConfig
	subscriptionConfigLock sync.RWMutex
)

func LoadSubscriptionConfig() error {
	absPath, err := filepath.Abs(plansConfigPath)
	if err != nil {
		return err
	}

	data, err := os.ReadFile(absPath)
	if err != nil {
		return err
	}

	var cfg SubscriptionConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return err
	}

	subscriptionConfigLock.Lock()
	subscriptionConfig = &cfg
	subscriptionConfigLock.Unlock()

	return nil
}

func GetSubscriptionConfig() *SubscriptionConfig {
	subscriptionConfigLock.RLock()
	defer subscriptionConfigLock.RUnlock()
	return subscriptionConfig
}

func StartSubscriptionConfigReloader() {
	ticker := time.NewTicker(configReloadInterval)
	defer ticker.Stop()

	for range ticker.C {
		if err := LoadSubscriptionConfig(); err != nil {
			log.Printf("ERROR: Failed to reload subscription config: %v", err)
		} else {
			log.Println("Subscription config reloaded successfully.")
		}
	}
}

func init() {
	if err := LoadSubscriptionConfig(); err != nil {
		log.Fatalf("FATAL: Failed to load initial subscription config: %v", err)
	}
}